Amazon1.Step4-Carousel

HTMLhead opened the pull request:
* 피드백반영 사항 
1. css함수 calc를 사용해 보려했는데 실패했습니다.. 쓸만한 곳을 찾아내기가 힘들더군요. 다른 곳에 사용해 볼 계획입니다.
2.  contentLoader.js와 topLayer.js는 합쳐서 어떤역할을 한다고 하신 말씀을 보고 여러가지로 생각해 보았습니다. 이벤트리스터에 들어가는 'contentLoaded'를 topLayer.js(지금은 scrollEvent-Sticky.js)파일 내부에 써볼까 생각했는데, 어짜피 나머지 js파일에도 contentLoaded가 들어가니까 contentLoader(지금은 app.js)에 한번에 모두 사용하는게 낫지않나 하는 생각에 이렇게 짜보았습니다.
3. entry point인 contentLoader.js 의 이름을 app.js 로 변경하였습니다. 이것이 더 entry point에 적합한 이름이라고 생각했습니다. 또한, 더 커지는 것을 방지하고, app.js 파일에는 어떤 일들이 벌어지는지만 간단히 명시하는 것이 좋다고 했던 토르코의 말이 기억나서 그렇게 해보려고 노력했습니다.

* 코드맥락
1. topLayer.js와 contentloader.js파일의 이름을 변경했습니다.(scrollEvent-Sticky.js, app.js)
1. module.js 파일을 만들어서 querySelector 와 ajax데이터를 가져오는 것을 조금 더 수월하게 할 수 있도록 만들어 보았습니다. 만들어놓은 class의 인자로 사용하도록 만들었는데, 이런식으로 사용하는게 맞는지 의구심이 들었습니다.
1. scrollEvent-Sticky.js파일에 intersectionObserver api를 사용해 보았습니다. yoon의 블로그를 참조하였습니다. 인자로 받은 header가 화면에서 사라지면 topLayer가 내려오고, 나타나면 다시 들어가도록 구현하였습니다.
1. 헤더말고도 body부분의 모든 css를 flex로 수정해보았습니다. 
1. 캐로샐 맥락
       1. 모든 이미지들이 들어가는 틀을 만들고 그 내부에 index0, 1, 2를 사용해서 배경이미지를 넣었습니다.
       2. ajax데이터로 배경이미지의 배열을 받아 만들고 그 배열의 값을 넣고 빼고 하며 무한으로 동작되도록 만들었습니다.
       3. 버튼의 방향으로 움직이는 애니메이션을 만들고 그 애니메이션이 끝나면 다시 원래대로 돌아오도록 만들었습니다. 하지만 배경이미지는 모두 바뀌어서 보기엔 오른쪽으로 넘어간 것처럼 느껴지도록 만들었습니다.
       4. 또한 애니메이션이 끝날 때 까지 버튼을 눌러도 동작되지 않도록 playBool을 추가하였습니다.
       5. 자동으로 동작되도록 하는 함수에는 setInterval을 사용하였습니다.
       6. 이미지를 모두 담아두고 있는 큰 틀에 mouseover와 mouseout이벤트를 등록하여 마우스가 큰 틀에 들어와 있을 때에는 동작하지 않도록 구현하였습니다.
       7. 이 동작을 하는 도중에는 mouseAuto 함수가 실행이 되는데, 여기에 isMouseOver불리언 값을 만들어서 만약 마우스가 안에 들어와 잇으면 자동으로 넘어가는 함수가 동작되지 않도록 만들었습니다.
1. init함수는 eventLister가 등록되는 함수들을 모아두고, domContent가 완료되었을 때 자동적으로 실행되어야 하는 동작들이 있는 함수는 constructor내부에서 실행되도록 구현하였습니다.

예제 페이지입니다.
[예제페이지](https://htmlhead.github.io/javascript-amazon/)

HTMLhead committed:
readme 업데이트

HTMLhead committed:
파일 이름 변경

contentLoader.js => app.js
topLayer.js => scrollEvent_Sticky.js

HTMLhead committed:
scrollEvent_sticky와 app파일수정

scrollEvent파일에서 해야하는 일을 app파일에서 해오고 있던 부분 수정

HTMLhead committed:
scrollEvent-sticky파일의 이름 변경

topLayer => stickyLayer, primeMember => hiddenLayer변수명
toplayerdrop => stickyLayerDrop, showprimemember => showHiddenLayer, removeLayer => hideLayer 함수명

HTMLhead committed:
버그수정

함수이름이 살짝달랐었음

HTMLhead committed:
scrollEvent_sticky에서 옵저버 패턴 사용중

HTMLhead committed:
쓸데없는 파일 제거 sticky쪽 스크롤 이벤트 수정

인터섹션옵저버를 사용하여 스크롤 이벤트 구현

HTMLhead committed:
scrollEvent_sticky.js 파일의 위치 수정

HTMLhead committed:
위치조정

HTMLhead committed:
topBody의 header쪽 폰트 수정

HTMLhead committed:
topBody.css파일정보 모두 css함수 사용하여 조정

HTMLhead committed:
sup태그 수정

HTMLhead committed:
carouselNav 수정

carousel-card모두 flex로 수정

HTMLhead committed:
carouselNav 부분 수정

위치잡는것 flex로 모두 수정하고 text-align:center사용해서 내부 컨텍스트들
모두 가운데로 수정, 마진값을 조정해서 index를 표현하는 반투명 구의 위치 수정

HTMLhead committed:
선택되었을 때의 index값수정

HTMLhead committed:
쓸데없는 백그라운드 이미지 maincarousel에서 모두 제거

HTMLhead committed:
main-carousel도 flex로 수정

HTMLhead committed:
mainCarousel.css파일 수정

버튼의 위치 flex로 지정, height를 주어 항상 가운데에 위치하도록 수정,
또한, carousel자체의 마진값을 바텀에 주어 좀떨어져 있을 수 있도록수정

HTMLhead committed:
middleBody carousel부분 수정

liststyle 제거

HTMLhead committed:
middleBodyCarousel파일 수정

flex를 이용하여 가운데로 정렬되도록 만든후 텍스트 위치 조정중

HTMLhead committed:
text부분 css완료

HTMLhead committed:
이미지 라인 태그와 css수정

HTMLhead committed:
middleBody.css수정

색깔들 모두 css함수 이용해서 수정

HTMLhead committed:
middleBodyCarousel.css파일 수정

주석 처리된 이전 css속성 모두 제거

HTMLhead committed:
module추가, json데이터 추가

querySelector를 간단화시키는 module파일 생성, 앞으로 ajax를 여러부분에 쓸것 같아서 ajax
를 부르는 기능도 module파일에 추가, backgroundImage를 모두 가지고 있는 json데이터를 추가
하였음.

HTMLhead committed:
middleBodyCarousel, common css파일 수정

common파일에 캐로셀 컨테이너의 크기를 정할수 있도록 저장, 그리고 적용

HTMLhead committed:
middleBodyCarousel.css수정

라인을 잡아주는 작업 시행중

HTMLhead committed:
middleBodyCarousel

사진의 위치를 잡아주는 작업 성공

HTMLhead committed:
middleBodyCarousel

이미지 위치에 opacity를 주어 자리를 차지하도록 수정

HTMLhead committed:
carousel_middle파일 생성, 스켈레톤코드 작성

HTMLhead committed:
carousel_middle.js 기능 추가

오른쪽으로 움직이게 하는 버튼의 기능 부여도중 css의 허점을 발견하고 css파일 수정

HTMLhead committed:
app부분에 Carousel_middle컨스트럭터에 들어갈 인자값 추가

HTMLhead committed:
오른쪽으로 움직이는 부분 완성

HTMLhead committed:
왼쪽으로 움직이는 부분 작성

HTMLhead committed:
button event에서 중복되는 부분 함수로 빼고 적용

HTMLhead committed:
carousel가리개 추가

추가하면서 이미지 변경

HTMLhead committed:
주석 수정

HTMLhead committed:
깃허브 페이지를 위한 이름수정

HTMLhead committed:
carousel_auto 파일추가

HTMLhead committed:
carousel_middle수정

carousel auto에 사용될 정보들이 carouselMiddle 에 너무많아서 carouselMiddle에
그냥 carousel auto기능까지 추가하도록 계획변경

HTMLhead committed:
carousel auto기능 추가

HTMLhead committed:
changeName

crongro commented:
제눈에는 슬라이딩 영역이 좌우측에 부분까지 일부분 보이고, 이부분이 슬라이딩 액션 이후에, 좀 늦게 로딩되서 보이는데,
버그 같습니다.
이부분수정할 수 있을까요?

crongro commented on amazonJS/app.js:
```diff
@@ -0,0 +1,20 @@
+    new Module()
+  );
+  const carousel_middle = new Carousel_middle(
+    false,
```
뭐가 false인지 알수가 없으니, 단일객체로 인자를 넣어주면 좋을 듯.
new Carousel_middle( {
  isGo : false,  listEl : document.querySelector(".middle-body-carousel-list')
});

이런식

crongro commented on amazonJS/app.js:
```diff
@@ -0,0 +1,20 @@
+import { Module } from "./module.js"
```
Module이라는 이름은 너무 일반적인 이름이라 안쓰는 게 좋을 듯.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  init() {
+    this.module.getAjax(this.handler.bind(this), `./jsonData/data.json`);
```
url도 옵션으로 볼 수 있을텐데요.
Carousel_middle 클래스 생성할때 주입받는 게 더 좋을 듯.
(클래스 재사용측면에서요)

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,100 @@
+    //롸잇, 레프트
+    const backgroundUrlArr = parsedObj.backgroundUrl;
+    for (let i = 0; i < backgroundUrlArr.length; i++) {
+      this.parsedArr.push(parsedObj["backgroundUrl"][i]);
```
배열을 순회하면서 push하는 거면 map을 쓸 수 있는것인지 검토바람.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,100 @@
+    this.playBool = true;
+    this.parsedArr.unshift(this.parsedArr.pop());
+    //shifting
+    this.container.style.transform = "translateX(230px)";
```
230px이런 하드코딩된 정보입력은 자제..이것도 계산에 의해서 뽑거나 함수나 클래스의 옵션으로 받거나.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,100 @@
+      "transitionend",
+      this.shuffleArr.bind(this)
+    );
+    this.container.style.transition = "all 0.5s";
```
all 0.5s 이런정보가 반복적인것이고, 변경이 되면 여기저기서 수정할 확률이 높음으로..
해결해보세요. ㅎ

crongro commented on amazonJS/module.js:
```diff
@@ -0,0 +1,17 @@
+class Module {
```
Module보다는 유용한 모음이면, 
util이 좀더 어울리는 이름일 듯.

crongro commented on amazonJS/scrollEvent-Sticky.js:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  makeIO(root) {
+    const io = new IntersectionObserver(entries => {
```
이건 질문 또는 고민꺼리인데요.
IntersectionObserver 가 지원하지 않는 브라우저에서는 어떻게동작될까요? 
브라우저 지원이 부족하다면 어떻게 하는게 좋을까요? 
(polyfill ?? 이런것이 해결책이 될지도...)

HTMLhead replied:
사파리에서 동작시키니까 아예 동작하질 않더군요. intersectionObserver를 발견하지 못했다면서 에러를 발환시켰습니다. flex도 잘 먹지 않는것 같구요....
오픈소스에 있는 polyfill을 가져와서 script태그에 넣어서 사용해 보았더니 intersectionObserver 이벤트는 잘 실행이 되더군요.
근데 이렇게 script를 하나 더 추가하면 이 script파일을 읽으면서 시간이 더 오래걸리지 않을까 이런 생각이 들었습니다. 현업에서 polyfill을 가져와서 이렇게 쓰기도 하나요? 그리고 그 polyfill에서 필요한 기능만 사용할 수 있도록 수정도 하는지 궁금합니다.

crongro reviewed (changes requested):
네 수고했어요, 
처음 UI개발이 좀 어려운데, 잘 고비를 넘고 있어요.
전체적으로 방향을 잘 잡아서 하고 잇네요.

몇개 리뷰남기부분 확인해서 수정해보세요.

기능완성도에서는 슬라이딩 될때 주변영역이 늦게 로딩되는 부분이아쉽고요.
그리고 슬라이딩방향이 일반적인 방향(좌/우 누를때 영역이동)인지 한번 다른 웹사이트와 비교해보세요,.

HTMLhead committed:
고정형으로 css변경

HTMLhead committed:
hider변경

크기가 늘어들든 줄어들든 계속해서 똑같은 위치를 가리도록 지정

HTMLhead committed:
scrollEvent_sticky부분의 인자를 모두 객체로 변경

HTMLhead committed:
module부분 import하는것 수정

파일마다 module을 import해서 module기능은 파일에서 실행하도록 진행, 그리하여 app을
좀더 가볍게 만들어 주도록 노력했음

HTMLhead committed:
scrollEvnet-sticky.js constructor수정

인자로받은 elObj에서 각각의 값을 가져와 값 설정

HTMLhead committed:
module 이름변경

HTMLhead committed:
util파일 변경

객체를 export하는게 아닌 함수들을 각자 export해서 파일마다 쓰이는 기능만 다로 export하도록
작성

HTMLhead committed:
app 의 carousel_middle부분 수정

이름들을 전부 클래스 이름만 가져오도록 수정

HTMLhead committed:
carousel-middle.js파일 constructor, init부분 수정

모두 qs로 가져오게 만들고, 자동으로 실행되어야 하는 함수들을 init함수에 몰아넣음

HTMLhead committed:
app, carousel-middle.js에서 url부분 수정

url을 인자로 받아서 ajax를 불러올때 사용하도록 수정

HTMLhead committed:
handler함수 수정

for문 map으로 수정, background가 아니라 imgUrl로 수정

HTMLhead committed:
중도과정 저장

HTMLhead committed:
middlebody carousel의 전체적 문제점 수정

background만 변경하는 작업을 했었는데, link도 바뀌어야 하는게 자연스럽다고 생각되어 html태그중
겉으로 보이는 것에 링크를 걸고, json데이터에 link데이터도 추가를 해주었음. 또한 ajax로 가져올 때
link배열도 가져오도록 변경, 가져올때 굳이 빈배열이 필요하지 않을것 같아제거

HTMLhead committed:
app.js에서 carousel_middle에 인자로받는 새로운객체 추가

option객체로 여러가지 옵션을 줄수 있도록 만듬

HTMLhead committed:
option추가, carousel-middle의 moveAuto, moveright수정

option에 시간과 transition할 조건들을 추가할수 있도록 수정하고, moveAuto와 moveRight에 적용.

HTMLhead committed:
intersectionObserver의 polyfill을 가져와 적용

HTMLhead committed:
overflow hidden을 잘 활용합시다!

HTMLhead committed:
carousel-middle.js 함수별 중복부분 나눔

moveright와 moveleft 함수 내의 transitionendEvent를 함수로 따로빼서 동작하도록 적용

HTMLhead committed:
carousel-middle 타이머 변경

setInterval을 setTimeout을 재귀적으로 사용할 수 있도록 변경

HTMLhead committed:
carousel-middle 주소도 바뀌도록 변경

버튼이 클릭되거나 자동으로 이동될 때, 그림에 맞게 주소도 변경되도록 수정

HTMLhead commented:
버그를 수정하였습니다.
1. 가리개의 크기와 overFlow : hidden을 좀더 적절히 이용해서 일부분이 보이는 부분을 수정하였습니다.
1. 슬라이딩 될때 주변영역이 늦게 로딩 된다고 말씀하신 부분이 어느부분인지를 몰라서 제대로 수정을 못했습니다만.. 이미지를 제대로 가리면 자연스레 해결되는 문제일것 같아서 그렇게 하였습니다.
1. 슬라이딩의 방향을 일반적으로 바꾸었습니다…ㅎㅎㅎㅎ 섬세함이 부족 했네요.

피드백해주신 부분을 반영해 보았습니다.
1. 단일객체로 인자를 넣었습니다.
1. module파일의 이름을 util로 수정하였습니다.
1. url또한 옵션으로 받을 수 있도록 작성하였습니다.
1. for문을 map으로 변경하였습니다.
1. 매직넘버들을 모두 단일객체를 사용한 인자로 받을 수 있도록 만들었습니다.
1. intersectionObserver의 polyfill을 script파일로 추가하였습니다.

나머지 개인적으로 마음에 들지 않는 부분을 수정하였습니다.
1. setInterval대신 재귀적 setTimeout을 사용하였습니다.
1. Json데이터에 주소를 가져오도록 수정하고, carousel container내부에 a태그를 넣어서 이미지가 바뀔 때 마다 주소 또한 바뀌도록 하는 이벤트를 추가했습니다.

피드백 부탁드립니다!!:sos:

crongro commented on amazonCss/app.css:
```diff
@@ -8,4 +8,5 @@
 @import url('./primeMember.css');
-@import url('./topLayer.css');
\ No newline at end of file
+@import url('./topLayer.css');
```
저도 이걸 잘 안 쓰긴하는데요.
성능적인 문제로 보통 안티패턴이라고 하긴해요. 참고해두세요.

https://www.keycdn.com/blog/css-performance

crongro commented on amazonCss/mainCarousel.css:
```diff
@@ -1,190 +1,122 @@
-    margin-bottom: 20px;
+  display: inline-block;
+  background-color: #00A8E1;
+  color: var(--color-basic);
```
var 잘 쓰고 있음~

crongro commented on amazonJS/app.js:
```diff
@@ -0,0 +1,22 @@
+import { ScrollEvent_sticky } from "./scrollEvent-Sticky.js";
+import { Carousel_middle } from "./carousel-middle.js";
+
+document.addEventListener("DOMContentLoaded", () => {
```
여기 이벤트핸들러가 길어졌네요. 지금은 괜찮고요.
좀더 길어지면 별도 함수로 분리하는 것도 생각하세요.

crongro commented on amazonJS/app.js:
```diff
@@ -0,0 +1,22 @@
+    stickyLayer: ".top-layer-container",
+    hiddenLayer: ".prime-member-container"
+  });
+  const carousel_middle = new Carousel_middle({
```
이렇게 풍부한 옵션을 기반으로 객체를 만드는 것은 범용성을 늘리는 것이라 아주 좋아요.

참고로, 여기 아래 options 을 기반으로 객체생성하는 거 보면 무척길죠~
https://github.com/nhnent/tui.tree/blob/production/docs/getting-started.md

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+    this.left = qs(elObj.leftBtn);
+    this.showingLinkEl = qs(elObj.showingLinkEl);
+    this.jsonUrl = urlObj.jsonUrl;
+    this.length = optionObj.movingLength;
```
좀더 구체적인 이름...필요.
this.length 가 무엇일지..

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+    this.right = qs(elObj.rightBtn);
+    this.left = qs(elObj.leftBtn);
+    this.showingLinkEl = qs(elObj.showingLinkEl);
+    this.jsonUrl = urlObj.jsonUrl;
```
json도 포맷에 대한 정의라서 이름으로 부적절.
의도와 맥락을 이름에서 드러내세요.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+    const imgUrlArr = parsedObj.backgroundUrl;
+    const linkUrlArr = parsedObj.linkArr;
+    this.imgUrlArr = imgUrlArr.map(v => v);
+    this.linkUrlArr = linkUrlArr.map(v => v);
```
이부분은 뭐에요? 
this.linkUrlArr = linkUrlArr;
이거랑 뭐가 다른지?

HTMLhead replied:
배열을 참조하도록 만드는게 아니라 새로 만들고 싶었는데 굳이 새로 만들 필요가 전혀 없었네요....ㅠㅠ 수정하겠습니다

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+  moveRight() {
+    if (this.playBool) return;
+    this.playBool = true;
+    this.imgUrlArr.push(this.imgUrlArr.shift());
```
배열조작을 해서 만드는게 좀 특이해보여요 ㅎ
괜찮습니다.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+    this.transitionendEvent();
+  }
+
+  initLink() {
```
이 함수가 linkUrlArr[1] 이값을 인자로 받으면 어떨까요? 
테스트용이한 코드에 대해서 고민해보세요.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  carouselAuto() {
+    setTimeout(this.moveAuto.bind(this), 3000);
```
3000  이런코드를 하드코딩하지 말고, 미리 선언해두고 그것을 사용하도록 하세요.

crongro commented on amazonJS/util.js:
```diff
@@ -0,0 +1,13 @@
+}
+
+export function qs(value) {
+  return document.querySelector(value);
```
querySelector 는 사실 document를 기준으로만 동작하는 건 아니죠.

div.querySelector.. 
이렇게 구현하면 기준점이 document라는 root노드가 아니라서 DOM 탐색이 좀더 빠르겠죠.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+    });
+    this.container.style.transition = "all 0s";
+    this.container.style.transform = "translateX(0px)";
+    this.playBool = false;
```
bPlay 라는 이름이 나을 듯.

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  moveAuto() {
+    if (this.isMouseOver) {
```
is는 동사니까 함수에쓰시고요.
b는 boolean 타입을 말해주니까
this.bMouseOver 가 더 좋을 듯.

crongro reviewed (changes requested):
수정잘 봤어요.
메서드 이름이 어느덧 동사+명사가 아니고, 애매한상태에요.
의도를 잘 드러내게 전체 메서드를 모두 수정해보세요.

그밖의 리뷰도 참고해주세요.

(화 이 팅)

HTMLhead committed:
css파일 모두 scss로 변경

HTMLhead committed:
폴더명 변경

HTMLhead committed:
app.js부분의 인자명 변경

HTMLhead committed:
util.js의 qs수정

인자를 두개로 두어 꼭 document가 아니어도 querySelector가 root노드가 아닌 값을
가질수 있도록 수정

HTMLhead committed:
scrollEvent-sticky.js의 qs수정

util의 바뀐 qs을 scrollEvent-sticky.js의 파일에 반영시키도록 수정

HTMLhead committed:
변경된 qs와 변수명 carousel-middle.js에 적용

HTMLhead committed:
carousel-middle.js의 변수명 수정, 옵션 추가

jsonUrl => ajaxDataUrl, carouselAutoMovingMS option추가

HTMLhead committed:
carousel-middle.js의 initLink수정

initLink함수에 현재 인자를 받아서 값을 수정할 수 있도록 변경

HTMLhead committed:
이전 커밋의 버그 수정

this.linkUrlArr[1]이라고 정확하게 수정, left와 right부분에 모두 잘 집어넣도록 함

HTMLhead committed:
carousel-middle.js의 bool값 수정

playBool => bPlay, isMouseOver => bMouseOver

HTMLhead committed:
오타수정

HTMLhead committed:
carousel-middle.js carouselauto와 qs부분 수정

carouselAuto가 굳이 init밖으로 나가서 실행되어야하는지 이유가 없어서 안으로 집어 넣음
qs에 굳이 document가 필요하지 않은 부분들 수정

HTMLhead commented:
* 피드백 주신 부분들을 수정해 보았습니다.
1. 크롱이 말씀해 주신 변수들이 모두 적합한것 같아서 그렇게 수정하였습니다.
1. 인자로 받아야 할 값들을 받고, 인자의 이름 또한 보자마자 어떤것을 뜻하는지 알 수 있도록 수정하려고 노력했습니다
1. css를 scss파일로 변경해보았습니다. 
scss의 문법은 하나도 쓰지 않았고, 그냥 한개의 app.css에 모든 css파일을 컴파일 해주고 싶어서 사용했습니다. 다음 스텝부터 찬찬히 scss에대해 공부해가며 하나씩 고쳐나갈 계획입니다.

피드백 부탁드립니다! 🆘

crongro commented on amazonJS/carousel-middle.js:
```diff
@@ -2,101 +2,97 @@ import { qs, getAjax } from "./util.js";
-    this.transitioning = optionObj.transitioning;
-    this.playBool = false;
-    this.isMouseOver = false;
+    this.container = qs(document, elObj.container);
```
constructor 에 값을 할당할때 Object.assign을 사용할 수 있어요.
이패턴이 편리해서 꽤 쓰이니 알아두세요.

crongro commented on amazonJS/scrollEvent-Sticky.js:
```diff
@@ -2,17 +2,17 @@ import { qs } from "./util.js";
-    this.hiddenLayer = qs(elObj.hiddenLayer);
+    this.header = qs(document, elObj.header);
+    this.stickyLayer = qs(document, elObj.stickyLayer);
+    this.hiddenLayer = qs(document, elObj.hiddenLayer);
```
이 코드보니 생각난건데요.
qs가 default parameter를 사용하면서 동작하게 할 수도 있을듯.(기본값은 document)

crongro commented on amazonScss/topBody.scss:
```diff
@@ -0,0 +1,72 @@
+  background-image:url(https://images-na.ssl-images-amazon.com/images/G/01/marketing/prime/detail_page/Header_01._CB513133335_.png);
+  background-color: #0079b4;
+  color: var(--color-basic);
+  height:396px;
```
다음 css 개발에서는 em/rem 중 하나를 골라서 사용해보세요.

crongro reviewed (approved):
잘 수정하셨어요.

scss는 좀더 공부해서 mixin이나 다양한 장점을 적용해보면 더 좋겠어요.

다음 미션 하셔도 될거 같아요. 

이번 미션 구현한 방식이 좀 특별한데요.
다른 사람이 애니메이션 어떻게 구현했는지 한번 물어서 이해해보세요.
